applications/assemblyscript-scale-codec.md

```diff
@@ -0,0 +1,58 @@
+
+## Project Description :page_facing_up: 
+
+SCALE is a lightweight codec for binary serialization and deserialization used in Substrate. Currently there are Rust, Python, Golang, C++ and JavaScript  implementations of the codec. The goal of the project is to deliver AssemblyScript implementation as a separate library. 
+
+The library will be required for encoding/decoding Polkadot Wasm Executor <> Wasm Runtime blob calls and more specifically:
+- Wasm Runtime blob compiled from AssemblyScript parsing runtime function calls from the Wasm executor.
+- Wasm runtime blob compiled from AssemblyScript calling the Polkadot Runtime Environment API (Host API).
+
+The library is a prerequisite for an AssemblyScript framework that generates runtimes or any runtime implemented in AssemblyScript loaded into a Polkadot Host.
+
+LimeChain is a blockchain-agnostic development company with a strong focus on developer tooling. We see Polkadot as an exciting technology and we hope to be able to help the developer community through various dev tools and implementations.
+## Team :busts_in_silhouette:
+
+* **Members:** Daniel Ivanov, Lyubomir Kiprov, Christian Vesselinov
+* **LinkedIn Profiles:** [Daniel](https://www.linkedin.com/in/daniel-k-ivanov/), [Lyubomir](https://www.linkedin.com/in/lyubomir-kiprov/), [Christian](https://www.linkedin.com/in/chris-veselinov/);
+* **Code Repos:** **[https://github.com/Daniel-K-Ivanov](https://github.com/Daniel-K-Ivanov); [https://github.com/bakasura980](https://github.com/bakasura980); [https://github.com/thcrnk](https://github.com/thcrnk)**
+* **Website:** **[https://limechain.tech](https://limechain.tech)**
+* **Legal Structure:** Legal Structure: LimeLabs Ltd., incorporated in Bulgaria, Dragan Tsankov 23A, 1113, Sofia, Bulgaria
+* **Team's Experience:** 
+Since 2017, LimeChain has worked on 50+ blockchain projects, including a strong track record of building developer tools for different protocols such as Ethereum, EOS, Aeternity and Corda. Some of the companies LimeChain has worked with are Celo, P&G, Raiffeisenbank, Status, Dapper Labs and Maker among others. The proposed developer team in particular also has experience with Substrate.
+
+## Development Roadmap :nut_and_bolt: 
+
+* **Total Estimated Duration:** 17 days
+* **Full-time equivalent (FTE):** 1.5
+* **Total Costs:** $13 304 (to be paid in the respective BTC amount on the day of payment)
+
+### Milestone 1
+* **Estimated Duration:** 17 days
+* **FTE:** 1.5
+* **Costs:** $13 304 (to be paid in the respective BTC amount on the day of payment)
+
+
+| Number | Deliverable | Specification | 
+| ------------- | ------------- | ------------- |
+| 1. | Implementing the Library |  Delivering a library that supports encoding/decoding the following types: fixed-width integers (signed and unsigned - 8, 16, 32, 64), bool, Big Int, Bool Array, Byte Array, Int Array, String Array, BigInt Array, Hash, Tuple. |  
```

Comment:
Big Integers are not supported in AssemblyScript as of now, but there's already a library by one of the AS core contributors for that https://github.com/MaxGraey/as-bignum. It possibly makes sense to build up on this. The same is done in the JS version of the SCALE codec with `bn.js`.
For the time estimate it might be important to know that only u128 is implemented with safe overflow/underflow traps.  There's only an unsafe implementation for u256 and none for i128 and i256 or anything bigger than that.

On that background, you should probably be more precise what's in scope / what's expected by the W3F / cc @Noc2

Reply:
> One approach may be to deliver an implementation that supports the basic types listed above and once it's required/requested

That sounds very reasonable! I just wanted to make sure that you're aware of the missing implementation.

Equally mature as the Rust implementation is the JS - or more precise, TypeScript - 
 implementation btw, which is also following the approach of having extendable types.  Unfortunately it's not possible to just transfer it to AssemblyScript, but might be worth to look at, if you haven't done so already.

It's mainly these two folders as part of the Polkadot JS API:
https://github.com/polkadot-js/api/tree/master/packages/types/src/codec 
https://github.com/polkadot-js/api/tree/master/packages/types/src/primitive

Reply:
I agree with Stefie. This sounds reasonable to me.